### PR TITLE
notification integration with socket

### LINKED ISSUE #96: FE: Notifications

### PR DESCRIPTION
**Pull request by jethoo**
### What this PR does (required):
- It is the integration of notifications with socket.
- Socket action for send-notification and receiving notification added in the backend
- create notification api service added in the frontend
- notifications sent real-time for messages sent and designs submitted for contest

### Screenshots / Videos (required):

https://user-images.githubusercontent.com/5088917/127945501-0cb0be90-428a-4f1c-b712-58c238f3e683.mov

https://user-images.githubusercontent.com/5088917/127945505-a0e6eba0-84d0-4e94-bb01-d4eb1a4b727e.mov

### Any information needed to test this feature (required):
- Post what steps to take to be able to test this feature (eg. "sign up as a new user and upload a profile picture from the navbar")

### Any issues with the current functionality (optional):
- List any problems where you were not able to complete all tasks on ticket
- List any problems this PR may cause for the project or other active PRs
- Post a screenshot/video of the problem, along with detailed console outputs where applicable

### CODE REVIEW

File: client/src/context/useSocketContext.tsx [hunk 1]
```diff
@@ -1,18 +1,22 @@
...
 
 export const SocketContext = createContext<ISocketContext>({
   socket: undefined,
   initSocket: () => null,
+  socketNotification: undefined,
 });
 
 export const SocketProvider: FunctionComponent = ({ children }): JSX.Element => {
   const [socket, setSocket] = useState<Socket | undefined>(undefined);
+  const [socketNotification, setSocketNotification] = useState<Notification>();
```
Comment (rajivtitus) [inline client/src/context/useSocketContext.tsx:19]: We already have a context for notifications, why are we storing them again here? :)

File: client/src/pages/Dashboard/Dashboard.tsx [hunk 2]
```diff
@@ -36,6 +36,14 @@ export default function Dashboard(): JSX.Element {
     initSocket();
   }, [initSocket]);
 
+  //sending users to the socket server
+  useEffect(() => {
```
Comment (rajivtitus) [inline client/src/pages/Dashboard/Dashboard.tsx]: As we discussed in our meeting today, maybe move this into the socket context along with any other socket events?

File: client/src/context/notificationContext.tsx [hunk 3]
```diff
@@ -26,6 +30,9 @@ export const NotificationProvider: FunctionComponent = ({ children }): JSX.Eleme
     getAll();
   }, []);
 
+  //notification coming from socket server 
+  socketNotification ? notifications?.push(socketNotification) : notifications;
```
Comment (rajivtitus) [inline client/src/context/notificationContext.tsx]: Please don't mutate state directly in React, use an updater function instead.

**Review by rajivtitus — COMMENTED** [review]
Good effort! Some comments, please have a look

File: client/src/context/useSocketContext.tsx (hunk 1 above)
Comment (jeewanbhusal) [reply client/src/context/useSocketContext.tsx:19]: This is the notification which is coming from the socket. I am storing it here and passing the 'socketNotification' to notificationContext. 
This was the only way I could get everything working.
I tried the other way around , getting the 'setNotifications' inside socketContext and updating the notifications. But that does not work. But this way everything functions.

File: client/src/pages/Dashboard/Dashboard.tsx (hunk 2 above)
Comment (jeewanbhusal) [reply client/src/pages/Dashboard/Dashboard.tsx]: I tried moving everything from Dashboard related to socket to socketContext., but I could not get it functioning.
I think the initSocket function needs to be called from the Dashboard. But I have removed sending user from the frontend to socket completely. In the next commit , which I am going to push., I have got it working using token. Therefore, users connected to socket will be taken care of by token. So, no need to send users from frontend after they are logged in.

File: client/src/context/notificationContext.tsx (hunk 3 above)
Comment (jeewanbhusal) [reply client/src/context/notificationContext.tsx]: sorry about that., with too much thinking going on with sockets., this completely went off my mind. I have made necessary changes in the next commit.

**Comment by jeewanbhusal** [comment]
> Good effort! Some comments, please have a look

I have made all necessary changes in the next commit I am about to push.

1. Token based socket works in the backend.
2. Users socket id is being tracked using token.
3. Next commit , which i am about to push., works completely with these practices.

File: client/src/context/useSocketContext.tsx (hunk 1 above)
Comment (rajivtitus) [reply client/src/context/useSocketContext.tsx:19]: For future reference, please bear in mind that we want to have only one single point of truth (having a notifications state in two different places can lead to use of inaccurate data)

**Review by rajivtitus — APPROVED** [review]

File: client/src/context/useSocketContext.tsx (hunk 1 above)
Comment (jeewanbhusal-1) [reply client/src/context/useSocketContext.tsx:19]: Thank you., I will certainly keep that in mind and also will try to solve this ., whenever i get time.

**Pull request merged**